### PR TITLE
Adopt Rust 2018

### PR DESCRIPTION
**Pull request by scholtzan**
Some `cargo fix` magic to adopt Rust 2018.

I did the following steps (in case someone wants to reproduce):
1. Add `#![feature(rust_2018_preview)]` to `lib.rs`
2. Run `cargo +nightly fix --allow-dirty --edition `
3. Remove `#![feature(rust_2018_preview)]`
4. Add `edition = '2018'` to `Cargo.toml`

## Related Issues
https://github.com/xi-editor/xi-editor/issues/1030

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

**Comment by russelltg** [comment]
You should not need the feature flag as rust 2018 is stable now. See https://rust-lang-nursery.github.io/edition-guide/editions/transitioning-an-existing-project-to-a-new-edition.html

**Comment by jansol** [comment]
@russelltg Yes but we generally try to stay compatible with a couple of versions older than the latest stable rust.

**Comment by raphlinus** [comment]
This bumps the Rust requirement to 1.31. With 1.30.1 I get `error: Edition 2018 is unstable and only available for nightly builds of rustc.`

That's not a "no," that's just pointing out that it varies from our usual policy. The `#![feature(rust_2018_preview)]` should be deleted and the README should be updated (it currently says 1.29).

**Comment by jansol** [comment]
CI also needs to be updated in that case.

**Comment by scholtzan** [comment]
Right, `#![feature(rust_2018_preview)]` would have not been necessary. I updated the version in `README`

**Comment by cmyr** [comment]
@scholtzan we specify the rust version in a bunch of places, I would just `grep '1.29.2' from the directory root (there will be at least the two CI scripts, maybe the `run_all_checks` script, etc)

**Comment by jansol** [comment]
Looks like some more `cargo fix` passes are in order, preferrably with the stable 1.31 toolchain.

File: rust/core-lib/src/view.rs [hunk 1]
```diff
@@ -11,25 +11,26 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+#![allow(clippy::range_plus_one)]
```
Comment (cmyr) [inline rust/core-lib/src/view.rs]: I would prefer to just fix whatever this was pointing to?

File: rust/lsp-lib/src/types.rs [hunk 2]
```diff
@@ -37,9 +36,6 @@ pub trait Callable: Send {
     );
 }
 
-//FIXME: this can be removed when this change makes it to stable:
-// https://github.com/rust-lang-nursery/rust-clippy/pull/3321
-#[cfg_attr(feature = "cargo-clippy", allow(boxed_local))]
```
Comment (cmyr) [inline rust/lsp-lib/src/types.rs]: 🥂

File: rust/rope/src/compare.rs [hunk 3]
```diff
@@ -13,9 +13,10 @@
 // limitations under the License.
 
 //! Fast comparison of rope regions, principally for diffing.
+#![allow(clippy::cast_ptr_alignment, clippy::unreadable_literal)]
```
Comment (cmyr) [inline rust/rope/src/compare.rs]: ditto keeping these attached to specific functions if possible; these lints are useful, and we only want to ignore them intentionally, case by case.

File: rust/rope/src/multiset.rs [hunk 4]
```diff
@@ -356,7 +356,7 @@ impl fmt::Debug for Subset {
                     '+'
                 };
                 for _ in 0..s.len {
-                    try!(write!(f, "{}", chr));
+                    r#try!(write!(f, "{}", chr));
```
Comment (cmyr) [inline rust/rope/src/multiset.rs]: we can just replace this with a `?` now I think

File: rust/core-lib/src/view.rs (hunk 1 above)
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: Yeah, there was one case where changing this would return `RangeInclusive` while `Range` is expected.

Comment (cmyr) [reply rust/core-lib/src/view.rs]: okay, cool.

Comment (scholtzan) [reply rust/core-lib/src/view.rs]: I'll open an issue for this once this PR gets merged. I think it's a beginner-friendly refactoring issue

**Pull request merged**